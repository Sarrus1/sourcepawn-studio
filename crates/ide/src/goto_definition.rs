use std::hash::Hash;

use base_db::FilePosition;
use hir::{HasSource, Semantics};

use line_index::TextRange;
use smol_str::{SmolStr, ToSmolStr};
use syntax::{utils::ts_range_to_text_range, TSKind};
use vfs::FileId;

use crate::{RangeInfo, RootDatabase};

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct NavigationTarget {
    pub name: SmolStr,
    pub file_id: FileId,
    pub full_range: TextRange,
    pub focus_range: Option<TextRange>,
}

impl NavigationTarget {
    pub fn focus_or_full_range(&self) -> TextRange {
        self.focus_range.unwrap_or(self.full_range)
    }
}

pub(crate) fn goto_definition(
    db: &RootDatabase,
    pos: FilePosition,
) -> Option<RangeInfo<Vec<NavigationTarget>>> {
    let sema = &Semantics::new(db);
    let preprocessing_results = sema.preprocess_file(pos.file_id);
    let tree = sema.parse(pos.file_id);
    let root_node = tree.root_node();

    if let Some((offset, def)) = sema.find_macro_def(&pos) {
        let file_id = def.file_id(sema.db);
        let u_range = offset.name_range();
        let source_tree = sema.parse(file_id);
        let name = def.name(db).map(|it| it.to_smolstr()).unwrap_or_default();
        let def_node = def.source(sema.db, &source_tree)?.value;
        let name_range = find_inner_name_range(&def_node);
        let target_preprocessing_results = sema.preprocess_file(file_id);

        let navs = vec![NavigationTarget {
            name,
            file_id,
            full_range: target_preprocessing_results
                .source_map()
                .closest_u_range_always(ts_range_to_text_range(&def_node.range())),
            focus_range: target_preprocessing_results
                .source_map()
                .closest_u_range_always(name_range)
                .into(),
        }];

        return RangeInfo::new(u_range, navs).into();
    }

    let offset: u32 = preprocessing_results
        .source_map()
        .closest_s_position_always(pos.offset)
        .into();

    let node = root_node.descendant_for_byte_range(offset as usize, offset as usize)?;
    let def = sema.find_def(pos.file_id, &node)?;
    let ts_range = ts_range_to_text_range(&node.range());
    let u_range = preprocessing_results
        .source_map()
        .closest_u_range_always(ts_range);

    let file_id = def.file_id(db);
    let source_tree = sema.parse(file_id);
    let name = def.name(db).map(|it| it.to_smolstr()).unwrap_or_default();
    let def_node = def.source(db, &source_tree)?.value;

    let name_range = find_inner_name_range(&def_node);

    let target_preprocessing_results = sema.preprocess_file(file_id);
    let navs = vec![NavigationTarget {
        name,
        file_id,
        full_range: target_preprocessing_results
            .source_map()
            .closest_u_range_always(ts_range_to_text_range(&def_node.range())),
        focus_range: target_preprocessing_results
            .source_map()
            .closest_u_range_always(name_range)
            .into(),
    }];

    RangeInfo::new(u_range, navs).into()
}

/// Find the range of the inner name node of a definition node if there is one.
/// Otherwise, return the range of the definition node.
pub fn find_inner_name_range(node: &tree_sitter::Node) -> TextRange {
    let name_range = match TSKind::from(node) {
        TSKind::methodmap_property_native | TSKind::methodmap_property_method => {
            node.children(&mut node.walk()).find_map(|child| {
                if matches!(
                    TSKind::from(child),
                    TSKind::methodmap_property_getter | TSKind::methodmap_property_setter
                ) {
                    Some(child.child_by_field_name("name")?.range())
                } else {
                    None
                }
            })
        }
        _ => node
            .child_by_field_name("name")
            .map(|name_node| name_node.range()),
    }
    .unwrap_or_else(|| node.range());

    ts_range_to_text_range(&name_range)
}
