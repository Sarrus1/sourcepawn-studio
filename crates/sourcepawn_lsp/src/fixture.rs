use anyhow::Result;
use crossbeam::channel::Receiver;
use itertools::Itertools;
use lsp_server::{Connection, Response};
use lsp_types::{
    notification::{DidOpenTextDocument, Exit, Initialized},
    request::ResolveCompletionItem,
    request::{Completion, Initialize, Shutdown},
    ClientCapabilities, CompletionItem, CompletionParams, CompletionResponse,
    DidOpenTextDocumentParams, InitializeParams, InitializedParams, Location, Position, Range,
    TextDocumentIdentifier, TextDocumentItem, TextDocumentPositionParams, Url, WorkspaceFolder,
};
use std::{
    env,
    fs::File,
    io,
    path::{Path, PathBuf},
    sync::Once,
    thread::JoinHandle,
    time::Duration,
};
use tempfile::{tempdir, TempDir};
use zip::ZipArchive;

use super::{LspClient, Server};
use store::options::Options;

#[derive(Debug)]
pub enum InternalMessage {
    OptionsRequested,
}

#[derive(Debug)]
pub struct Fixture {
    pub documents: Vec<Document>,
}

impl Fixture {
    pub fn parse(input: &str) -> Fixture {
        let mut documents = Vec::new();

        let mut start = 0;
        if !input.is_empty() {
            for end in input
                .match_indices("%!")
                .skip(1)
                .map(|(i, _)| i)
                .chain(std::iter::once(input.len()))
            {
                documents.push(Document::parse(&input[start..end]));
                start = end;
            }
        }

        Self { documents }
    }

    pub fn write_files(&self, dir: &Path) {
        for document in &self.documents {
            let text = String::from(&document.text);
            let path = dir.join(&document.path);
            std::fs::create_dir_all(path.parent().unwrap()).unwrap();
            std::fs::write(&path, &text).unwrap();
        }
    }

    pub fn setup(&self, client: &LspClient, dir: &Path) {
        for document in &self.documents {
            let text = String::from(&document.text);
            let path = dir.join(&document.path);

            let uri = Url::from_file_path(&path).unwrap();
            let language_id = "sourcepawn".to_string();

            client
                .send_notification::<DidOpenTextDocument>(DidOpenTextDocumentParams {
                    text_document: TextDocumentItem::new(uri, language_id, 0, text),
                })
                .unwrap();
        }
    }
}

#[derive(Debug)]
pub struct Document {
    pub path: PathBuf,
    pub text: String,
    pub cursor: Option<Position>,
    pub ranges: Vec<Range>,
}

impl Document {
    pub fn parse(input: &str) -> Self {
        let mut lines = Vec::new();

        let (path, input) = input
            .trim()
            .strip_prefix("%! ")
            .map(|input| input.split_once('\n').unwrap_or((input, "")))
            .unwrap();

        let mut ranges = Vec::new();
        let mut cursor = None;

        for line in input.lines() {
            if line.chars().all(|c| matches!(c, ' ' | '^' | '|' | '!')) && !line.is_empty() {
                let index = (lines.len() - 1) as u32;

                cursor = cursor.or_else(|| {
                    let character = line.find('|')?;
                    Some(Position::new(index, character as u32))
                });

                if let Some(start) = line.find('!') {
                    let position = Position::new(index, start as u32);
                    ranges.push(Range::new(position, position));
                }

                if let Some(start) = line.find('^') {
                    let end = line.rfind('^').unwrap() + 1;
                    ranges.push(Range::new(
                        Position::new(index, start as u32),
                        Position::new(index, end as u32),
                    ));
                }
            } else {
                lines.push(line);
            }
        }

        Self {
            path: PathBuf::from(path),
            text: lines.join("\n"),
            cursor,
            ranges,
        }
    }
}

static LOGGER: Once = Once::new();

#[derive(Debug)]
pub struct TestBed {
    fixture: Fixture,
    locations: Vec<Location>,
    _temp_dir: TempDir,
    temp_dir_path: PathBuf,
    pub internal_rx: Receiver<InternalMessage>,
    client: LspClient,
    client_thread: Option<JoinHandle<()>>,
    server_thread: Option<JoinHandle<()>>,
}

impl Drop for TestBed {
    fn drop(&mut self) {
        log::trace!("Dropping server.");
        let _ = self.client.send_request::<Shutdown>(());
        let _ = self.client.send_notification::<Exit>(());
        self.client_thread.take().unwrap().join().unwrap();
        self.server_thread.take().unwrap().join().unwrap();
    }
}

impl TestBed {
    pub fn new(fixture: &str, add_sourcemod: bool) -> Result<Self> {
        LOGGER.call_once(|| {
            if option_env!("TEST_LOG") == Some("1") {
                fern::Dispatch::new()
                    .level(log::LevelFilter::Trace)
                    .chain(std::io::stderr())
                    .apply()
                    .unwrap()
            }
        });

        let fixture = Fixture::parse(fixture);

        let temp_dir = tempdir()?;
        let temp_dir_path = temp_dir.path().canonicalize()?;

        let temp_sm_dir = tempdir()?;
        let temp_sm_dir_path = temp_sm_dir.path().canonicalize()?;

        let locations: Vec<Location> = fixture
            .documents
            .iter()
            .flat_map(|document| {
                let uri = Url::from_file_path(temp_dir_path.join(&document.path)).unwrap();
                document
                    .ranges
                    .iter()
                    .map(move |range| Location::new(uri.clone(), *range))
            })
            .collect();

        let (server_conn, client_conn) = Connection::memory();
        let (internal_tx, internal_rx) = crossbeam::channel::unbounded();

        let client = LspClient::new(client_conn.sender);

        let server_thread =
            std::thread::spawn(move || Server::new(server_conn, false).run().unwrap());
        let client_thread = {
            let client = client.clone();
            std::thread::spawn(move || {
                let destination = temp_sm_dir_path.clone();
                for message in &client_conn.receiver {
                    match message {
                        lsp_server::Message::Request(request) => {
                            if request.method == "workspace/configuration" {
                                let mut options = Options::default();
                                if add_sourcemod {
                                    let current_dir = env::current_dir().unwrap();
                                    let sourcemod_path =
                                        current_dir.join("test_data/sourcemod.zip");
                                    unzip_file(&sourcemod_path, destination.to_str().unwrap())
                                        .unwrap();
                                    options
                                        .includes_directories
                                        .push(destination.clone().join("include/"));
                                }
                                client
                                    .send_response(Response::new_ok(request.id, vec![options]))
                                    .unwrap();
                                std::thread::sleep(Duration::from_millis(100));
                                internal_tx.send(InternalMessage::OptionsRequested).unwrap()
                            } else {
                                client
                                    .send_error(
                                        request.id,
                                        lsp_server::ErrorCode::MethodNotFound,
                                        "Method not found".into(),
                                    )
                                    .unwrap();
                            }
                        }
                        lsp_server::Message::Response(response) => {
                            client.recv_response(response).unwrap();
                        }
                        lsp_server::Message::Notification(_) => {}
                    }
                }
            })
        };
        fixture.write_files(&temp_dir_path);

        Ok(TestBed {
            fixture,
            locations,
            _temp_dir: temp_dir,
            temp_dir_path,
            client,
            internal_rx,
            client_thread: Some(client_thread),
            server_thread: Some(server_thread),
        })
    }

    pub fn initialize(&self, capabilities: ClientCapabilities) -> Result<()> {
        self.client
            .send_request::<Initialize>(InitializeParams {
                capabilities,
                workspace_folders: Some(vec![WorkspaceFolder {
                    uri: Url::from_file_path(self.temp_dir_path.clone()).unwrap(),
                    name: "test".to_string(),
                }]),
                ..Default::default()
            })
            .unwrap();
        self.client
            .send_notification::<Initialized>(InitializedParams {})
            .unwrap();

        self.fixture.setup(&self.client, self.directory());

        Ok(())
    }

    pub fn client(&self) -> &LspClient {
        &self.client
    }

    pub fn cursor(&self) -> Option<TextDocumentPositionParams> {
        let (document, cursor) = self
            .fixture
            .documents
            .iter()
            .find_map(|document| document.cursor.map(|cursor| (document, cursor)))?;

        let uri = Url::from_file_path(self.temp_dir_path.join(&document.path)).unwrap();
        let id = TextDocumentIdentifier::new(uri);
        Some(TextDocumentPositionParams::new(id, cursor))
    }

    #[allow(unused)]
    pub fn locations(&self) -> &[Location] {
        &self.locations
    }

    pub fn directory(&self) -> &Path {
        &self.temp_dir_path
    }

    #[allow(unused)]
    pub fn documents(&self) -> &[Document] {
        &self.fixture.documents
    }
}

pub fn complete(fixture: &str) -> Vec<CompletionItem> {
    let test_bed = TestBed::new(fixture, true).unwrap();
    test_bed
        .initialize(
            serde_json::from_value(serde_json::json!({
                "textDocument": {
                    "completion": {
                        "completionItem": {
                            "documentationFormat": ["plaintext", "markdown"]
                        }
                    }
                },
                "workspace": {
                    "configuration": true,
                    "workspace_folders": true
                }
            }))
            .unwrap(),
        )
        .unwrap();
    let text_document_position = test_bed.cursor().unwrap();
    test_bed
        .internal_rx
        .recv_timeout(Duration::from_secs(10))
        .unwrap();
    let items = match test_bed
        .client()
        .send_request::<Completion>(CompletionParams {
            text_document_position,
            partial_result_params: Default::default(),
            work_done_progress_params: Default::default(),
            context: None,
        })
        .unwrap()
    {
        Some(CompletionResponse::Array(items)) => items,
        Some(CompletionResponse::List(list)) => list.items,
        None => Vec::new(),
    };

    items
        .into_iter()
        .map(|item| match item.data {
            Some(_) => {
                let mut item = test_bed
                    .client()
                    .send_request::<ResolveCompletionItem>(item)
                    .unwrap();

                item.data = None;
                item
            }
            None => item,
        })
        .sorted_by(|item1, item2| item1.label.cmp(&item2.label))
        .collect()
}

pub fn unzip_file(zip_file_path: &PathBuf, destination: &str) -> Result<(), io::Error> {
    let file = File::open(zip_file_path)?;
    let mut archive = ZipArchive::new(file)?;

    for i in 0..archive.len() {
        let mut file = archive.by_index(i)?;
        let dest_path = format!("{}/{}", destination, file.name());

        if file.is_dir() {
            std::fs::create_dir_all(&dest_path)?;
        } else {
            if let Some(parent_dir) = std::path::Path::new(&dest_path).parent() {
                std::fs::create_dir_all(parent_dir)?;
            }
            let mut dest_file = File::create(&dest_path)?;
            io::copy(&mut file, &mut dest_file)?;
        }
    }

    Ok(())
}
