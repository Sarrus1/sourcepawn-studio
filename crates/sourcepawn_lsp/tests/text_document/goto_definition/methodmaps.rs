use insta::assert_json_snapshot;
use sourcepawn_lsp::fixture::goto_definition;

#[test]
fn methodmap_1() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {}
           |
           ^
"#,
    ));
}

#[test]
fn methodmap_2() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {}

methodmap Bar < Foo {}
                 |
                 ^
"#,
    ));
}

#[test]
fn methodmap_method_1() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public void Foo() {}
                 |
                 ^
}
"#,
    ));
}

#[test]
fn methodmap_method_2() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public void Foo(int foo) {}
                         |
                         ^
}
"#,
    ));
}

#[test]
fn methodmap_method_3() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public void Foo(int foo) {
        foo += 1;
         |
         ^
    }
}
"#,
    ));
}

#[test]
fn methodmap_method_4() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public void Foo() {}
    public void Bar() {
        this.Foo();
          |
          ^
    }
}
"#,
    ));
}

#[test]
fn methodmap_method_5() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public void Foo() {}
    public void Bar() {
        this.Foo();
              |
              ^
    }
}
"#,
    ));
}

#[test]
fn methodmap_method_6() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public void Foo() {}
}

Foo foo;

void main() {
    foo.Foo();
         |
         ^
}
"#,
    ));
}

#[test]
fn methodmap_method_7() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public void Foo() {int bar}
}

Foo foo;

void main() {
    int bar;
    foo.Foo(bar);
             |
             ^
}
"#,
    ));
}

#[test]
fn methodmap_native_method_1() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public native Foo();
                   |
                   ^
}
"#,
    ));
}

#[test]
fn methodmap_native_method_2() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public native Foo(int foo);
                           |
                           ^
}
"#,
    ));
}

#[test]
fn methodmap_constructor_1() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public Foo() {}
            |
            ^
}
"#,
    ));
}

#[test]
fn methodmap_constructor_2() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public Foo(int foo) {}
                    |
                    ^
}
"#,
    ));
}

#[test]
fn methodmap_constructor_3() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public Foo(int foo) {
        foo += 1;
         |
         ^
    }
}
"#,
    ));
}

#[test]
fn methodmap_destructor_1() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public ~Foo() {}
             |
             ^
}
"#,
    ));
}

#[test]
fn methodmap_property_1() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {}
                  |
                  ^
}
"#,
    ));
}

#[test]
fn methodmap_property_2() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
enum Bar {
    Baz
}

methodmap Foo {
    property Bar Foo {}
              |
              ^
}
"#,
    ));
}

#[test]
fn methodmap_property_3() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {}
}

Foo foo;

void main() {
    foo.Foo;
         |
         ^
}
"#,
    ));
}

#[test]
fn methodmap_property_getter_1() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {
        public get() { return 1; }
                |
                ^
    }
}
"#,
    ));
}

#[test]
fn methodmap_property_getter_2() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
int foo;
methodmap Foo {
    property int Foo {
        public get() { 
            return foo;
                    |
                    ^
        }
    }
}
"#,
    ));
}

#[test]
fn methodmap_property_getter_3() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {
        public get() { 
            int foo;
            return foo;
                    |
                    ^
        }
    }
}
"#,
    ));
}

#[test]
fn methodmap_property_setter_1() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {
        public set(int foo) {}
                |
                ^
    }
}
"#,
    ));
}

#[test]
fn methodmap_property_setter_2() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {
        public set(int foo) {}
                        |
                        ^
    }
}
"#,
    ));
}

#[test]
fn methodmap_property_setter_3() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {
        public set(int foo) {
            foo += 1;
             |
             ^
        }
    }
}
"#,
    ));
}

#[test]
fn methodmap_property_setter_4() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {
        public set(int foo) {
            int foo;
            foo += 1;
             |
             ^
        }
    }
}
"#,
    ));
}

#[test]
fn methodmap_property_native_1() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {
        public native get();
                       |
                       ^
        }
    }
}
"#,
    ));
}

#[test]
fn methodmap_property_native_2() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {
        public native set(int foo);
                       |
                       ^
        }
    }
}
"#,
    ));
}

#[test]
fn methodmap_property_native_3() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    property int Foo {
        public native set(int foo);
                               |
                               ^
        }
    }
}
"#,
    ));
}

#[test]
fn methodmap_new_1() {
    assert_json_snapshot!(goto_definition(
        r#"
%! main.sp
methodmap Foo {
    public Foo() {}
    public void bar() {}
}

void main() {
    Foo foo = new Foo();
    foo.bar();
         |
         ^
}
"#,
    ));
}
