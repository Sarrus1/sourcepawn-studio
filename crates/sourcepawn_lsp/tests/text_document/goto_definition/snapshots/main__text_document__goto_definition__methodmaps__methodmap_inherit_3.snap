---
source: crates/sourcepawn_lsp/tests/text_document/goto_definition/methodmaps.rs
assertion_line: 553
expression: "goto_definition(r#\"\n%! main.sp\nmethodmap Foo {\n    property int Foo\n    {\n        public get() {}\n        public set(int value) {}\n    }\n    property int Bar\n    {\n        public get() {}\n        public set(int value) {}\n    }\n}\nmethodmap Bar < Foo {\n    property int Bar\n    {\n        public get() {}\n        public set(int value) {}\n    }\n}\nBar bar;\nvoid main() {\n    bar.Bar;\n         |\n         ^\n}\n\"#)"
---
[
  {
    "originSelectionRange": {
      "start": {
        "line": 21,
        "character": 8
      },
      "end": {
        "line": 21,
        "character": 11
      }
    },
    "targetUri": "file:///main.sp",
    "targetRange": {
      "start": {
        "line": 13,
        "character": 4
      },
      "end": {
        "line": 17,
        "character": 5
      }
    },
    "targetSelectionRange": {
      "start": {
        "line": 13,
        "character": 17
      },
      "end": {
        "line": 13,
        "character": 20
      }
    }
  }
]
